src/pip/_internal/commands/install.py

```diff
@@ -315,7 +321,21 @@ def run(self, options, args):
                         ignore_installed=options.ignore_installed,
                         isolated=options.isolated_mode,
                     )
+                    top_level_reqs = list(requirement_set.requirements.keys())
                     resolver.resolve(requirement_set)
+                    if options.deps_only:
+                        if options.editables:
+                            raise CommandError("Cannot use --deps-only "
+                                               "with editable installations")
+                        if options.requirements:
+                            raise CommandError("Cannot use --deps-only with "
+                                               "requirements files")
+
+                        for target in top_level_reqs:
+                            del requirement_set.requirements[target]
```

Comment:
No time to do a thorough review, but won't this have the following impact (I'll do this with "pure" requirements, not with extras, just to simplify the argument):

Suppose you have `A` and `B` and B depends on A. Then `pip install --deps-only B` will install A, but `pip install --deps-only A B` won't. Meaning that we *haven't* installed the dependencies of B, like the user asked us to.

It also means that
```
pip install --deps-only A B
```
and
```
pip install --deps-only A
pip install --deps-only B
```
will behave differently, which is possibly confusing.

I'm not sure there's an "obvious" best answer to this, but (a) I think the behaviour should be clearly specified (and that means there's probably a need for a docs patch here as well) and (b) I think the logic of what "dependencies only" should mean needs to be more clearly thought through for cases like this that maybe aren't the "intended" use of the feature (#4783 is the intended use) but which may crop up once people have the chance to use the feature.

It's possible that the behaviour as implemented here is "good enough", so all of the above is theoretical and not worth worrying about - but again, let's be explicit rather than simply leaving it assumed.

Reply:
Good point ! Maybe a `--fake-install pkg` option (allowing multiple packages) would provide better semantics..

Reply:
I thought about this while I worked on the patch and decided to go with the simplest solution just because this seemed the least surprising behaviour. I think it would be more surprising if ‘—deps-only’ would proceed to install requirements just because they are dependencies as well. Nevertheless, I’m happy to add another commit to document whichever solution we agree on.

Reply:
@pfmoore: I pushed a commit to document the currently implemented behaviour. Let me know what you think.